File: 0303-range-sum-query-immutable/0303-range-sum-query-immutable.js
/**
 * @param {number[]} nums
 */
var NumArray = function(nums) {
    this.cumulativeSum = [0];
    
    for (let i = 0; i < nums.length; i++) {
        this.cumulativeSum[i + 1] = this.cumulativeSum[i] + nums[i];
    }
};

/** 
 * @param {number} left 
 * @param {number} right
 * @return {number}
 */
NumArray.prototype.sumRange = function(left, right) {
    return this.cumulativeSum[right + 1] - this.cumulativeSum[left];
};

/** 
 * Your NumArray object will be instantiated and called as such:
 * var obj = new NumArray(nums)
 * var param_1 = obj.sumRange(left,right)
 */

File: 0303-range-sum-query-immutable/0303-range-sum-query-immutable.py
class NumArray(object):

    def __init__(self, nums):
        """
        :type nums: List[int]
        """
        self.nums = nums
        self.cum_sum = 0
        self.prefix_sum = []
        for num in self.nums:
            self.cum_sum += num
            self.prefix_sum.append(self.cum_sum)
        
        

    def sumRange(self, left, right):
        """
        :type left: int
        :type right: int
        :rtype: int
        """
        if left == 0:
            return self.prefix_sum[right] - 0
        else:
            return self.prefix_sum[right] - self.prefix_sum[left - 1]
        


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# param_1 = obj.sumRange(left,right)